### DISCUSSION TITLE
Calculate current day, direction

### DISCUSSION THREAD

**Discussion by 2075** [cosinekitty/astronomy#336] [Q&A]
Hi there!

I found your awesome library and use it in a Kotlin application and have some questions, maybe you are able to direct me towards a solution. Sorry in advance for my noob questions.

1. I would like to calculate the total solar days and solar days passed in a year of a planet ( ```planetOrbitalPeriod```?), where a day is full rotation of itself and a year a full orbit around the sun. Is this possible to achieve with the current lib? I saw some references in the docs, but nothing like a function call like getTotalDays(body) or getCurrentDay(body). Atm I believe I have to manually calculate based on pericenter/apocenter results of the `searchPlanetApsis` result?

2. I would like to calculate the direction of a planet from a given local Observer position on earth. Is `horizontal` the correct approach to retrieve the direction in degrees and the alltitude?

```
val time = Time.fromMillisecondsSince1970(Calendar.getInstance().timeInMillis)
val observer = Observer( latLonElevNull.first , latLonElevNull.second, latLonElevNull.third )
val equatorial = equator(it, timeA, observer, EquatorEpoch.OfDate, Aberration.Corrected)
val horizontal = horizon(time, observer, equatorial.ra, equatorial.dec, Refraction.Normal)
```

3. Another question is regarding spin in AxisInfo. It does not seem normalized to 0-360 degrees, but the documentation refers to it as the rotation in degrees, but I get e.g. ```D  Earth        3187718.7326```. Would you mind to explain?

Thanks in advance @cosinekitty  :)!

**Comment by cosinekitty** [comment]
Hi @2075! Sorry it took so long to get back to you. My work life has been bonkers lately!

One of the issues you will find is that ideas like "years" and "days" are intuitive but imprecise. For example, planets only "sort of" orbit the Sun. In practice, the Sun is wobbling around too, mostly being tugged on by Jupiter and Saturn. As planets move around in the solar system, they make an approximate ellipse shape that never quite matches up where it was last time. Thus there are different ways to define a planet's year. Which definition is "right" is an opinion, not a fact of nature.

For example, take the planet Neptune. If you ask when it is at perihelion, it turns out to be a messy thing because the Sun wobbles around faster than Neptune orbits. This means there are several times in Neptune's year when it is "closest" to the Sun, in terms of having a locally minimum distance. 

Therefore, without knowing what your real goal/motivation is, I would guess you will get better results using ecliptic longitude (the angle around the Sun) as your measuring stick for deciding when a new orbit/year begins.

I don't currently have a function that searches for when a planet crosses a given ecliptic longitude. However, I do have the [eclipticLongitude](https://github.com/cosinekitty/astronomy/blob/master/source/kotlin/doc/ecliptic-longitude.md) function, which will tell you what the ecliptic longitude of a planet is around the Sun's center. You may be able to use that to search for when the planet crosses a given longitude. A good choice would be 180 degrees, since that is as far as possible as the ugly discontinuity from 359.999... degrees to 0 degrees.

So that would take care of planetary years.

About a planet's days... this is simpler, and then you have to ask whether you want solar days or sidereal days. On Earth, what we commonly call a day is a solar day, the time it takes for the Sun to roughly reach the same position in the sky, averaged over an entire year.

A sidereal day is how long it takes for a distant star to come back to the same position in the observer's sky. The difference is a planet can orbit the Sun by a significant angle over one of its rotations, thus making solar and sidereal days different. For example, on Earth a sidereal day is about 4 minutes shorter than a solar day. The extra 4 minutes is how long it takes the Earth to rotate slightly more than 360 degrees so that it once again faces the new direction of the Sun.

If you take a look at the code for [rotationAxis](https://github.com/cosinekitty/astronomy/blob/61dc07020aaa6885d2c7f688a4d82beaf6edb9ef/source/kotlin/src/main/kotlin/io/github/cosinekitty/astronomy/astronomy.kt#L4546), you can see that the `spin` angle is quite complicated to calculate in general. I think it makes sense to call rotationAxis at two different times (the beginning and end of a single year/orbit), subtract the angles, and divide by 360. This will give you sidereal days. You can approximate a solar day by subtracting exactly one day (one rotation) from the number of rotations per orbit to obtain a good estimate of the correction.

I hope this helps!

**Reply by 2075** [reply]
Hey @cosinekitty, wow, that is a comprehensive answer, thank you for that. I think with the described approaches for `eclipticLongitude` and `rotationAxis` I will get exactly what I need. 

In general the usage of these two is more for illustrative purposes inside a "Solar System Watchface" which I am working on. As the use in real space for such an app is unlikely, and the approach is more to educate about the differences in time, the ecliptic longitude might be enough for getting a grasp on a planetary year.

Regarding SOL or sidereal, I am currently working with solar days, and the calulation suggested seems a bit simpler than spin :) thanks. I am not 100% confident, I am using the right approach to get the local rotation on a planet, e.g. when I compare rotation of Saturn vs Venus and rendering it, there should be notable differences in time progression on the watchface.

Will ping with more info and possibly more questions.

Again thanks very much for the great lib and also your personal dedication!
